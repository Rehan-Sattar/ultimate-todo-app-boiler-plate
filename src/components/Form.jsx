import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { insertTodoToDatabase } from "../store/ActionCreators";
import "../App.css";
class Form extends Component {
    constructor() {
        super();
        this.state = {
            title: '',
            description: '',
            doneStatus: false,
        };
        this.handleSubmitForm = this.handleSubmitForm.bind(this);
    };

    handleSubmitForm(event) {
        event.preventDefault();
        this.props.inserTodo(this.state);
        this.setState({
            title: '',
            description: '',
            doneStatus: false,
        });
    };
    render() {
        return (
            <div>
                <div className="container">
                    <div className="row justify-content-center">
                        <div className="col-md-8 col-lg-8 col-sm-12" id="form-container">
                            <h2 className="text-center" id="addTaskHeadline"> <i className="fas fa-pencil-alt"></i> Add your task</h2>
                            <form onSubmit={this.handleSubmitForm} className="p-3">
                                <input
                                    type="text"
                                    placeholder="Title"
                                    required
                                    className="form-control"
                                    value={this.state.title}
                                    onChange={(e) => this.setState({ title: e.target.value })} />
                                <br />
                                <textarea
                                    placeholder="Description"
                                    required
                                    className="form-control"
                                    value={this.state.description}
                                    onChange={(e) => this.setState({ description: e.target.value })} >
                                </textarea>
                                <br />
                                <button type={"Submit"} className="btn cstmBtn btn-block btn-lg"> <i className="fa fa-plus"></i> Add Task</button>
                            </form>
                        </div>
                    </div>
                </div>

            </div>
        );
    };
};

const mapStateToProps = () => ({})
const mapDispatchToProps = (dispatch) => bindActionCreators({
    inserTodo: (Todostate) => insertTodoToDatabase(Todostate)
},
    dispatch);
export default connect(mapStateToProps, mapDispatchToProps)(Form);