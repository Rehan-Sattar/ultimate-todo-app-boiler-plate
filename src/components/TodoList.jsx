import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import Modal from "react-responsive-modal";
import {
    getAllTodosFromDatabase,
    deleterTodoFromDatabase,
    updateTodoInDatabase,
    taskDoneAttempt
} from "../store/ActionCreators";

class TodoList extends Component {
    constructor() {
        super();
        this.state = {
            todos: [],
            open: false,
            updateTitle: '',
            updateDescription: '',
            updateDoneStatus: false,

            addAction: false

        };
        this.handleTaskdone = this.handleTaskdone.bind(this);
        this.handleDeleteRequest = this.handleDeleteRequest.bind(this);
        this.handleUpdateRequest = this.handleUpdateRequest.bind(this);
    };

    onOpenModal = (id) => {
        this.setState({ open: true });
        localStorage.setItem('updatedTodoId', id);
    };

    onCloseModal = () => {
        this.setState({ open: false });
    };

    componentDidMount() {
        this.props.downloadTodos();
    }

    componentWillReceiveProps(props) {
        this.setState({
            todos: props.todos.todos,
            addAction: props.addedTodoAction.addAction
        });
    };

    handleDeleteRequest(todoId) {
        this.props.deleteATask(todoId);
    }


    handleUpdateRequest(e) {
        e.preventDefault();
        const { updateDescription, updateTitle, updateDoneStatus } = this.state;
        this.props.updateTodoTask({
            updateDescription,
            updateTitle,
            updateDoneStatus,
            todoId: localStorage.getItem('updatedTodoId')
        });
        this.onCloseModal();
    }

    handleTaskdone(todo, status) {
        this.props.changeDoneStatus(todo, status);
    }
    render() {
        const { open } = this.state;
        return (
            <div>
                <div className="container">
                    <div className="row justify-content-center">
                        <div className="col-md-10 col-lg-10 col-sm-12">
                            {this.state.todos ? this.state.todos.map((todo, index) => {
                                if (!todo.done) {
                                    return (
                                        <div className="card my-5" key={index}>
                                            <div className="card-body">
                                                <h3 className="todotext">
                                                    {todo.title}
                                                </h3>
                                                <p className="todotext">
                                                    {todo.description}
                                                </p>
                                                <div className="btn-group float-right">
                                                    <button
                                                        onClick={this.handleTaskdone(todo.id, false)}
                                                        className="btn btn-outline-success btn-lg card_btns"> <i className="fa fa-check"></i> </button>
                                                    <button
                                                        onClick={() => this.onOpenModal(todo.id)}
                                                        className="btn btn-outline-warning btn-lg card_btns"> <i className="fas fa-pencil-alt"></i> </button>
                                                    <button
                                                        onClick={() => this.handleDeleteRequest(todo.id)}
                                                        className="btn btn-outline-danger btn-lg card_btns"> <i className="fa fa-trash"></i> </button>
                                                </div>
                                            </div>
                                        </div>
                                    )
                                } else {
                                    return (
                                        <div className="card my-5 doneBody" key={index}>
                                            <div className="card-body">
                                                <h3 className="todotext crossed">
                                                    {todo.title}
                                                </h3>
                                                <p className="todotext crossed">
                                                    {todo.description}
                                                </p>
                                                <div className="btn-group float-right">
                                                    <button
                                                        onClick={this.handleTaskdone(todo.id, false)}
                                                        className="btn btn-outline-warning btn-lg card_btns"> <i className="fas fa-undo"></i> </button>

                                                    <button
                                                        onClick={() => this.handleDeleteRequest(todo.id)}
                                                        className="btn btn-outline-danger btn-lg card_btns"> <i className="fa fa-trash"></i> </button>
                                                </div>
                                            </div>
                                        </div>
                                    )
                                }
                            }) : <p>No items</p>}
                        </div>
                    </div >
                </div >

                <Modal open={open} onClose={this.onCloseModal} center>
                    <h2 className="text-center my-2 pt-2">Update Todo</h2>
                    <form onSubmit={this.handleUpdateRequest}>
                        <input
                            type="text"
                            placeholder="Title"
                            required
                            className="form-control"
                            onChange={(e) => this.setState({ updateTitle: e.target.value })} />
                        <br />
                        <br />
                        <textarea
                            placeholder="Description"
                            required
                            className="form-control"
                            onChange={(e) => this.setState({ updateDescription: e.target.value })} >
                        </textarea>
                        <br />
                        <br />
                        <button className="btn btn-primary btn-block" type={"Submit"}>Update</button>
                    </form>
                </Modal>
            </div >
        );
    };
};

const mapStateToProps = (state) => {
    return {
        todos: state.allTodos,
        addedTodoAction: state.addedTodo
    }
}
const mapDispatchToProps = (dispatch) => bindActionCreators({
    downloadTodos: (Todostate) => getAllTodosFromDatabase(Todostate),
    deleteATask: (todoId) => deleterTodoFromDatabase(todoId),
    updateTodoTask: (newTodoObject) => updateTodoInDatabase(newTodoObject),
    changeDoneStatus: (todo) => taskDoneAttempt(todo)
},
    dispatch);
export default connect(mapStateToProps, mapDispatchToProps)(TodoList);