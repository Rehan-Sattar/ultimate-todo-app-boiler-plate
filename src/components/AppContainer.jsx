import React, { Component } from "react";
import Form from "./Form";
import List from "./TodoList";
import "../App.css";
class AppContainer extends Component {
    render() {
        return (
            <div>
                <Form />
                <List />
            </div>
        );
    };
};

export default AppContainer;