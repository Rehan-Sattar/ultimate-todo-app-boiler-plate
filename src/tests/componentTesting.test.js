import React from "react";
import { shallow } from "enzyme";
import "../SetupTest";
import { Provider } from "react-redux"
import configureStore from "redux-mock-store";
import AppContainer from "../components/AppContainer";
import Form from "../components/Form";
import TodoList from "../components/TodoList";

const mockStore = configureStore();
const store = mockStore({
    allTodos: [],
    error: '',
    addedTodo: false
})
// shallow Rendering of AppContainer

it('should render the AppContainer Component', () => {
    const wrapper = shallow(<AppContainer />);
    expect(wrapper.length).toEqual(1);
});


it('should check the nodes | child Components', () => {
    const wrapper = shallow(<AppContainer />);
    expect(wrapper.contains([
        <Form />,
        <TodoList />
    ])
    )
});


// form container shallow rendering
it('should render the Form component', () => {
    const wrapper = shallow(
        <Provider store={store}>
            <Form />
        </Provider>
    );
    expect(wrapper.length).toEqual(1);
});


// snapshot testing

it('should render and snap the AppContainer Component | snapshot testing', () => {
    const wrapper = shallow(<AppContainer />);
    expect(wrapper).toMatchSnapshot();
});

it('should render and snap the Form Component | snapshot testing', () => {
    const wrapper = shallow(
        <Provider store={store}>
            <Form />
        </Provider>);
    expect(wrapper).toMatchSnapshot();
});

it('should render and snap the TodoList Component | snapshot testing', () => {
    const wrapper = shallow(<Provider store={store}>
        <TodoList />
    </Provider>);
    expect(wrapper).toMatchSnapshot();
});

