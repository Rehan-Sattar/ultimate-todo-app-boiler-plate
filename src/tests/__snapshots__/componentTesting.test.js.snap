// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render and snap the AppContainer Component | snapshot testing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(Form) />,
        <Connect(TodoList) />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(Form) />,
          <Connect(TodoList) />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`should render and snap the Form Component | snapshot testing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
>
    <Connect(Form) />
</Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`should render and snap the TodoList Component | snapshot testing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
>
    <Connect(TodoList) />
</Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
